beancount/parser/lexer.l

```diff
@@ -301,15 +309,13 @@ ACCOUNTNAME     ([A-Z0-9]|{UTF-8-ONLY})([A-Za-z0-9\-]|{UTF-8-ONLY})*
     beancount.core.flags, in order for round-trips between text and entries to
     be possible. {5307d8fa1e7b}
     */
-[!&#?%PSTCURM]	{
-    if (yy_line_tokens != 1) {
-        yylval->character = yytext[0];
-        return FLAG;
-    }
-    else {
-        yy_skip_line();
-        return SKIPPED;
-    }
+<AFTERDATE>[!&#?%PSTCURM]  {
+    yylval->character = yytext[0];
+    return FLAG;
+}
+^[!&#?%PSTCURM]    {
+    yy_skip_line();
+    return SKIPPED;
 }
```

Comment:
I like this. But I think `*` should be added to the flag characters. Also, it requires adjusting the grammar too: in a few places looks for an `HASH` or `ASTERISK` token (see the `txn` grammar rule).

Reply:
Looking into this, `*`, `#`, and `:` were all handled separately before this line, with their own beginning-of-line checks. In fact, it's not clear to me that the previous code *ever* triggered for `#`.  Adding all three to this arrangement seems like it would be fine: I don't think there are places where any of them are used after a DATE, or after an INDENT, to mean anything other than a FLAG.  Then rules for separate HASH, ASTERISK, and COLON tokens can be moved to after this handling, with the separate beginning-of-line check removed.

However, I'm a bit uncomfortable here, because looking through the grammar, it appears that HASH is actually used in what appears to be an undocumented but implemented compound cost statement, and ASTERISK is used in an undocumented and apparently not yet implemented cost merging statement.  While I think my new changes preserve these behaviours, I'm hoping that they don't break any other potential behaviour I'm not aware of.

Reply:
Hmm... there seems to be a wider problem here.  `beancount.core.flags` has `[*!PSTCURM]` as valid flags, as I now notice you pointed out in #516.

Reply:
"#" is used in compound costs, to specify unit and total cost, as {unit # total currency}.

Reply:
Yes.  Before this change, "#" was tokenized early on as HASH if not the first token, and otherwise the line was skipped.  "#" does not appear to ever trigger line 304, as it is caught by line 257.  In the grammar, it can be parsed as a flag, but that doesn't fit with beancount.core.flags (see also https://github.com/beancount/beancount/issues/516).  

After the change, "#" is never parsed as a flag, per beancount.core.flags.  The line-skipping and "#" in a compound cost is not affected by the change.  Alternatively, "#" could be added to line 275 to be parsed as a flag, without affecting either of hose.